### DISCUSSION TITLE
Avoiding Address Reuse

### DISCUSSION THREAD

**Discussion by InteXX** [btcpayserver/btcpayserver#5263] [General]
I'm new to crypto—in fact, I haven't processed my first transaction—so bear with me please if I seem a bit slow on the uptake.

As I understand things, it's good security practice to not reuse wallet addresses. As such, I'd like to automatically create a new address for each incoming transaction (I'm going to be accepting monthly subscriptions at my website; members will be paying for single or multiple months in advance).

I just watched [this video](https://youtu.be/MIWGvl6_WzI), and I believe I understand it to say that a given wallet will be publicly visible. It is this I wish to avoid (again, for security purposes). If necessary to avoid it, I'm prepared to create a new wallet for every transaction—as long as I'm able to somehow fully automate the task.

Is what I'm describing making any sense for the BTCPay world?

In another (non-BTCPay) angle down this security path, ChatGPT suggests this code to generate multiple keys for the same wallet (I'm a VB.NET programmer):
```vb
Imports NBitcoin

Module Program
    Sub Main()
        ' Create a master key for the HD wallet
        Dim masterExtKey As ExtKey = New ExtKey()

        ' Derive child key pairs
        Dim childKey1 As ExtKey = masterExtKey.Derive(New KeyPath("0/0"))
        Dim childKey2 As ExtKey = masterExtKey.Derive(New KeyPath("0/1"))

        ' Get corresponding public keys
        Dim publicKey1 As PubKey = childKey1.Neuter().PubKey
        Dim publicKey2 As PubKey = childKey2.Neuter().PubKey

        Console.WriteLine("Public Key 1: " & publicKey1.ToString())
        Console.WriteLine("Public Key 2: " & publicKey2.ToString())
    End Sub
End Module
```
That said, given my infancy in all of this, I'm not sure I trust myself to be able to build a secure application for this purpose.

Ideas? Suggestions? Corrections?

**Comment by Kukks** [comment]
BTCPay Server handles this internally automatically. Please try to use it first. Your code is unrelated to how to use BTCPay Server.

**Comment by InteXX** [comment]
That sounds good, and thank you, but the narrator on the [previously linked video](https://youtu.be/MIWGvl6_WzI?t=150) seems to be stating otherwise:

> The first thing you need to do is add a derivation scheme to your wallet. This is the public key of your wallet. Whoever has access to this key cannot access your funds, but they can see all the addresses that your wallet has.

This would seem to make a unique-address-per-transaction approach pointless. An attacker can discover the amount of funds in my wallet and thereby determine whether he wishes to attempt a break-in.

Could you clarify? Am I misunderstanding something?

**Comment by InteXX** [comment]
Am I off-topic with this one?

Some discussion one way or another sure would help, thanks.
